### ISSUE TITLE
Allow factory delegates to return null

### ISSUE THREAD

**Issue by slipdef** [autofac/Autofac#708]
`builder.Register(c => c.Resolve<IXXXProvider>().Get()).As<IXXX>();`
IXXXProvider returns null in some cases. Autofac throws "A delegate registered to create instances of 'IXXX' returned null". I don't see any clear reason to don't allow resolving dependencies in nulls if custom factory method is used.

**Comment by tillig** [comment]
I think there are a couple of reasons this is disallowed:

1. **People don't expect to get null when they resolve a service.** Given the majority IoC pattern is to pass _required_ dependencies into a consuming component via the constructor, allowing a resolution operation of any kind to return null breaks a kind of contract with that consuming component. Instead of being able to assume the required dependency is passed in, consumers now have to handle errors and "what if?" scenarios.
1. **From a DI perspective, the client doesn't know whether it's a factory or a reflection creation.** From a flexibility standpoint, the component requiring the dependency doesn't know that you registered using a factory method - you just as easily could have registered using reflection, a situation where you'll never get null when you new-up an object. Having all of the resolution mechanisms work in a standard way means you have behavior you can count on - the "contract" I mentioned earlier.

From a precedent perspective, this is actually how many IoC containers also work. For example, if you tried to do this in Ninject...

```csharp
public class CarModule : NinjectModule
{
  public override void Load()
  {
    Bind<IXXXProvider>().To<XXXProvider>();
    Bind<XXX>().ToMethod(ctx => ctx.Kernel.Get<IXXXProvider>().Get());
  }
}
```

When you call `kernel.Get<XXX>()` to invoke your provider, on returning null you'd get an error like this:

```
[ActivationException] Error activating XXX using binding from XXX to method
Provider returned null.
Activation path:
  1) Request for XXX

Suggestions:
  1) Ensure that the provider handles creation requests properly.
```

In a nutshell: this is functioning as designed.
